src/main/scala/ir/transforms/Inline.scala

```diff
@@ -0,0 +1,162 @@
+  case Unreachable(label) => EventuallyUnreachable(label)
+  case Return(label, out) =>
+    EventuallyReturn(
+      out.toList.map { case y: (Variable, Expr) =>
+        ((y._1.name, visit_expr(varName, y._2)))
+      }.toArray,
+      label
+    )
+}
+
+def keyToString[T](varRenamer: CILVisitor)(x: (Variable, Expr)): (String, Expr) =
+  (x(0).name, visit_expr(varRenamer, x(1)))
+
+def convertStatementRenaming(varRenamer: CILVisitor)(x: Statement): EventuallyStatement = x match {
+  case DirectCall(targ, outs, actuals, label) =>
+    directCall(
+      outs.toArray.map(v => (v(0).name, visit_rvar(varRenamer, v(1)))),
+      targ.name,
+      actuals.toArray.map(keyToString(varRenamer)): _*
+    )
+  case IndirectCall(targ, label) => indirectCall(visit_rvar(varRenamer, targ))
+  case x: NonCallStatement =>
+    CloneableStatement(visit_stmt(varRenamer, cloneStatement(x)).head.asInstanceOf[NonCallStatement])
+}
+
+def convertBlockRenaming(varRenamer: CILVisitor, blockName: String => String)(x: Block) = {
+  EventuallyBlock(
+    blockName(x.label),
+    x.statements.toArray.map(convertStatementRenaming(varRenamer)),
+    convertJumpRenaming(blockName, varRenamer, x.jump),
+    x.address
+  )
+}
+
+/**
+  * Inline a procedure call to the calling procedure;
+  *
+  * - maintains param form 
+  * - maintain dsa form if both procedures are in dsa form
+  *
+  * require invariant.SingleCallBlockEnd
+  */
+def inlineCall(prog: Program, c: DirectCall): Unit = {
+  require(c.target.entryBlock.isDefined)
+  require(c.target.returnBlock.isDefined)
+
+  val proc = c.parent.parent
+  val block = c.parent
+  val target = c.target
+  // rename ssa variables to maintain DSA in the new procedure
+  // rename block labels to avoid creating conflicts if inlining the same function multiple times
+  val renamer = VarRenamer(proc)
+  val blockRenamer: String => String = memoise(renameBlock)
+
+  val entry = target.entryBlock.get
+  val returnBlock = target.returnBlock.get
+
+  // the internal blocks to the target
+  val internalBlocks = (target.blocks.toSet -- Seq(entry, returnBlock)).map(convertBlockRenaming(renamer, blockRenamer))
+
+  // clone the target procedure and apply the renaming
+  val (entryTempBlock, entryResolver) = convertBlockRenaming(renamer, blockRenamer)(entry).makeResolver
+  val eventuallyReturnBlock = convertBlockRenaming(renamer, blockRenamer)(returnBlock)
+  val afterCallBlock = block.createBlockAfter("_inlineret")
+  proc.addBlock(entryTempBlock)
+  block.replaceJump(GoTo(entryTempBlock))
+
+  // replace return in target with a jump to the aftercall block
+  val (returnTemp, resolveReturnBlock) = eventuallyReturnBlock.copy(j = unreachable).makeResolver
+  proc.addBlock(returnTemp)
+
+  // resolve internal call blocks
+  val resolvers = internalBlocks.map(_.makeResolver)
+  resolvers.foreach { case (block, _) => proc.addBlock(block) }
+  resolvers.foreach { case (_, resolve) => resolve(prog, proc) }
```

Comment:
i _think_ you might be able to use EventuallyBlock.resolve here instead of going through makeResolver
```suggestion
  internalBlocks.foreach(b => proc.addBlock(b.resolve(prog, proc)))
```

Reply:
I doubt this will work if the procedure contains a loop the resolve on a goto will fail the target isn't already a member of the procedure its resolving against.